### PR TITLE
[CDAP-12382] Properties mapping from triggering pipeline to arguments in triggered pipeline

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-12382
https://builds.cask.co/browse/CDAP-RUT1307

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 1]
```diff
@@ -94,6 +100,8 @@
 public class SmartWorkflow extends AbstractWorkflow {
   public static final String NAME = "DataPipelineWorkflow";
   public static final String DESCRIPTION = "Data Pipeline Workflow";
+  public static final String TRIGGERING_PROPERTIES_MAPPING = "triggeringPropertiesMapping";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: on the design doc this is triggering.properties.mapping

**Comment by ecfm** [comment]
The first commit is the main change and the second commit is to remove unnecessary DefaultMacroEvaluator constructors as mentioned in #9421

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+  @VisibleForTesting
+  public void addPluginPropertyMapping(PluginPropertyMapping pluginPropertyMapping) {
+    pluginProperties.add(pluginPropertyMapping);
+  }
+
+  /**
+   * The mapping between a triggering pipeline argument to a triggered pipeline argument.
+   */
+  public static class ArgumentMapping {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java]: make these their own classes, not inner classes

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  /**
+   * @return The list of mapping between triggering pipeline plugin properties to triggered pipeline arguments
+   */
+  public List<PluginPropertyMapping> getPluginProperties() {
+    return pluginProperties;
+  }
+
+  @VisibleForTesting
+  public void addArgumentMapping(ArgumentMapping argumentMapping) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java]: avoid state change when possible. Remove these adders and just pass in maps in a constructor

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * The mapping between triggering pipeline properties to the triggered pipeline arguments.
+ */
+public class TriggeringPipelinePropertyMapping {
+  private final List<ArgumentMapping> arguments;
+  private final List<PluginPropertyMapping> pluginProperties;
+
+  public TriggeringPipelinePropertyMapping() {
+    this.arguments = new ArrayList<>();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java]: use Collections.emptyList() so that it's unmodifiable

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    public ArgumentMapping(String source, @Nullable String target) {
+      this.source = source;
+      this.target = target;
+    }
+
+    /**
+     * @return The name of triggering pipeline argument
+     */
+    public String getSource() {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/TriggeringPipelinePropertyMapping.java]: this is actually nullable too, due to user error, since it's created through deserialization.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java [hunk 6]
```diff
@@ -40,15 +41,13 @@
...
                                long logicalStartTime, SecureStore secureStore, String namespace) {
     Map<String, String> resolvedArguments = new HashMap<>();
-    resolvedArguments.putAll(runtimeArguments);
-    // not expected, but can happen if user runs just the program and not the workflow
-    if (workflowToken != null) {
-      for (String tokenKey : workflowToken.getAll(WorkflowToken.Scope.USER).keySet()) {
-        resolvedArguments.put(tokenKey, workflowToken.get(tokenKey, WorkflowToken.Scope.USER).toString());
-      }
+    Iterator<Map.Entry<String, String>> iter = arguments.iterator();
+    while (iter.hasNext()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java]: don't put things into the map. Just store the arguments and use that class instead of the map.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java [hunk 7]
```diff
@@ -109,9 +109,8 @@ public MapReduceTransformExecutorFactory(MapReduceTaskContext taskContext,
     this.metrics = metrics;
     this.sourceStageName = sourceStageName;
     this.macroEvaluator =
-      new DefaultMacroEvaluator(taskContext.getWorkflowToken(), taskContext.getRuntimeArguments(),
-                                taskContext.getLogicalStartTime(), taskContext,
-                                taskContext.getNamespace());
+      new DefaultMacroEvaluator(new BasicArguments(taskContext.getWorkflowToken(), taskContext.getRuntimeArguments()),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java]: don't create a new one here, it's already passed into the constructor

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java [hunk 8]
```diff
@@ -128,10 +127,9 @@ private MapReduceRuntimeContext createRuntimeContext(StageSpec stageInfo) {
   private <IN, ERROR> TrackedMultiOutputTransform<IN, ERROR> getMultiOutputTransform(StageSpec stageSpec)
     throws Exception {
     String stageName = stageSpec.getName();
-    DefaultMacroEvaluator macroEvaluator = new DefaultMacroEvaluator(taskContext.getWorkflowToken(),
-                                                                     taskContext.getRuntimeArguments(),
-                                                                     taskContext.getLogicalStartTime(), taskContext,
-                                                                     taskContext.getNamespace());
+    DefaultMacroEvaluator macroEvaluator =
+      new DefaultMacroEvaluator(new BasicArguments(taskContext.getWorkflowToken(), taskContext.getRuntimeArguments()),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java]: use the arguments already in the class

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java [hunk 9]
```diff
@@ -147,10 +145,9 @@ private MapReduceRuntimeContext createRuntimeContext(StageSpec stageInfo) {
   @SuppressWarnings("unchecked")
   private <IN, OUT> TrackedTransform<IN, OUT> getTransformation(StageSpec stageSpec) throws Exception {
 
-    DefaultMacroEvaluator macroEvaluator = new DefaultMacroEvaluator(taskContext.getWorkflowToken(),
-                                                                     taskContext.getRuntimeArguments(),
-                                                                     taskContext.getLogicalStartTime(), taskContext,
-                                                                     taskContext.getNamespace());
+    DefaultMacroEvaluator macroEvaluator =
+      new DefaultMacroEvaluator(new BasicArguments(taskContext.getWorkflowToken(), taskContext.getRuntimeArguments()),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java]: same here

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 10]
```diff
@@ -198,7 +199,7 @@ public void initialize() throws Exception {
     hConf.setBoolean("mapreduce.reduce.speculative", false);
 
     // plugin name -> runtime args for that plugin
-    MacroEvaluator evaluator = new DefaultMacroEvaluator(context.getWorkflowToken(), context.getRuntimeArguments(),
+    MacroEvaluator evaluator = new DefaultMacroEvaluator(new BasicArguments(context),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: pipelineRuntime.getArguments()

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/PipelineAction.java [hunk 11]
```diff
@@ -81,8 +82,8 @@ public void run() throws Exception {
                                                             phaseSpec.isProcessTimingEnabled());
     Action action =
       pluginContext.newPluginInstance(stageSpec.getName(),
-                                      new DefaultMacroEvaluator(context.getWorkflowToken(),
-                                                                context.getRuntimeArguments(),
+                                      new DefaultMacroEvaluator(new BasicArguments(context.getWorkflowToken(),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/PipelineAction.java]: move PipelineRuntime above this line and get it from pipelineRuntime

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java [hunk 12]
```diff
@@ -84,8 +85,9 @@ public boolean apply(@Nullable WorkflowContext input) {
     PluginContext pluginContext = new PipelinePluginContext(input, metrics, phaseSpec.isStageLoggingEnabled(),
                                                             phaseSpec.isProcessTimingEnabled());
 
-    MacroEvaluator macroEvaluator = new DefaultMacroEvaluator(input.getToken(), input.getRuntimeArguments(),
-                                                              input.getLogicalStartTime(), input, input.getNamespace());
+    MacroEvaluator macroEvaluator =
+      new DefaultMacroEvaluator(new BasicArguments(input.getToken(), input.getRuntimeArguments()),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java:89]: get it from pipelineRuntime

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 13]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
     super.initialize(context);
-
-    String arguments = Joiner.on(", ").withKeyValueSeparator("=").join(context.getRuntimeArguments());
+    TriggeringScheduleInfo scheduleInfo = context.getTriggeringScheduleInfo();
+    if (scheduleInfo != null) {
+      String propertiesMappingString = scheduleInfo.getProperties().get(TRIGGERING_PROPERTIES_MAPPING);
+      if (propertiesMappingString != null) {
+        TriggeringPipelinePropertyMapping propertiesMapping =
+          GSON.fromJson(propertiesMappingString, TriggeringPipelinePropertyMapping.class);
+        Map<String, String> newRuntimeArgs = getNewRuntimeArgsFromScheduleInfo(scheduleInfo, propertiesMapping);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: there's no need to get this map and then traverse the map again and put everything in the token. Instead, set things in the token as you go.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 14]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    addTriggeringPipelineArguments(propertiesMapping.getArguments(), triggerInfo, newRuntimeArgs);
+    addTriggeringPipelinePluginProperties(propertiesMapping.getPluginProperties(), triggerInfo, newRuntimeArgs);
+    return newRuntimeArgs;
+  }
+
+  private void addTriggeringPipelineArguments(List<TriggeringPipelinePropertyMapping.ArgumentMapping> argumentMappings,
+                                              ProgramStatusTriggerInfo triggerInfo,
+                                              Map<String, String> runtimeArgs) {
+    Map<String, String> triggerRuntimeArgs = triggerInfo.getRuntimeArguments();
+    if (triggerRuntimeArgs == null) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: triggerInfo returns null, not an empty map? Doesn't seem right, should change it if it does. In any case, you should be checking if it's empty here.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 15]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+      if (!argumentMappings.isEmpty()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: nested ifs should be avoided if possible. This should be a single if condition.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 16]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+        LOG.warn("No runtime arguments in the run '{}' of the triggering pipeline '{}' in namespace '{}' ",
+                 triggerInfo.getRunId(), triggerInfo.getApplicationSpecification().getName(),
+                 triggerInfo.getNamespace());
+      }
+      return;
+    }
+    for (TriggeringPipelinePropertyMapping.ArgumentMapping mapping : argumentMappings) {
+      String value = triggerRuntimeArgs.get(mapping.getSource());
+      if (value == null) {
+        if (!argumentMappings.isEmpty()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: this is always true

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 17]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    return newRuntimeArgs;
+  }
+
+  private void addTriggeringPipelineArguments(List<TriggeringPipelinePropertyMapping.ArgumentMapping> argumentMappings,
+                                              ProgramStatusTriggerInfo triggerInfo,
+                                              Map<String, String> runtimeArgs) {
+    Map<String, String> triggerRuntimeArgs = triggerInfo.getRuntimeArguments();
+    if (triggerRuntimeArgs == null) {
+      if (!argumentMappings.isEmpty()) {
+        LOG.warn("No runtime arguments in the run '{}' of the triggering pipeline '{}' in namespace '{}' ",
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: 'in the run' -> 'in run'

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 18]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+                 triggerInfo.getRunId(), triggerInfo.getApplicationSpecification().getName(),
+                 triggerInfo.getNamespace());
+      }
+      return;
+    }
+    for (TriggeringPipelinePropertyMapping.ArgumentMapping mapping : argumentMappings) {
+      String value = triggerRuntimeArgs.get(mapping.getSource());
+      if (value == null) {
+        if (!argumentMappings.isEmpty()) {
+          LOG.warn("Runtime argument '{}' is not found in the run '{}' of the triggering pipeline '{}' " +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: 'in the run' -> 'in run'

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 19]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    if (triggerRuntimeArgs == null) {
+      if (!argumentMappings.isEmpty()) {
+        LOG.warn("No runtime arguments in the run '{}' of the triggering pipeline '{}' in namespace '{}' ",
+                 triggerInfo.getRunId(), triggerInfo.getApplicationSpecification().getName(),
+                 triggerInfo.getNamespace());
+      }
+      return;
+    }
+    for (TriggeringPipelinePropertyMapping.ArgumentMapping mapping : argumentMappings) {
+      String value = triggerRuntimeArgs.get(mapping.getSource());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: you need to handle the case where getSource() returns null because a user gave you bad data. In that case, skip the mapping and continue

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 20]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    }
+    // Currently only expecting one trigger in a schedule
+    ProgramStatusTriggerInfo triggerInfo = programStatusTriggerInfos.get(0);
+    // Get the value of every triggering pipeline property and put it to newRuntimeArgs if it's not null
+    addTriggeringPipelineArguments(propertiesMapping.getArguments(), triggerInfo, newRuntimeArgs);
+    addTriggeringPipelinePluginProperties(propertiesMapping.getPluginProperties(), triggerInfo, newRuntimeArgs);
+    return newRuntimeArgs;
+  }
+
+  private void addTriggeringPipelineArguments(List<TriggeringPipelinePropertyMapping.ArgumentMapping> argumentMappings,
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: there isn't any value in having this as a separate method since it's not re-used. It just breaks up the logical flow and makes it harder to read.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java (hunk 19 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: it's also not correct to just read the runtime arguments. Like we discussed, you need to be reading the workflow token in order to read arguments from pipelines higher up than the direct parent. You should be creating a BasicArguments class using the triggering pipeline's runtime arguments and workflow token and using that class here.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 21]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+          LOG.warn("Runtime argument '{}' is not found in the run '{}' of the triggering pipeline '{}' " +
+                     "in namespace '{}' ",
+                   mapping.getSource(), triggerInfo.getRunId(), triggerInfo.getApplicationSpecification().getName(),
+                   triggerInfo.getNamespace());
+        }
+        continue;
+      }
+      // Use the argument name in the triggering pipeline if target is not specified
+      String key = mapping.getTarget() == null ? mapping.getSource() : mapping.getTarget();
+      runtimeArgs.put(key, value);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: you can just put it directly into the token here

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 22]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    }
+  }
+
+  private void addTriggeringPipelinePluginProperties(
+    List<TriggeringPipelinePropertyMapping.PluginPropertyMapping> pluginMapping,
+    ProgramStatusTriggerInfo triggerInfo, Map<String, String> runtimeArgs) {
+    Map<String, Plugin> plugins = triggerInfo.getApplicationSpecification().getPlugins();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: note: this should be changed once https://github.com/caskdata/cdap/pull/9485 is merged.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 23]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    Map<String, String> triggerRuntimeArgs = triggerInfo.getRuntimeArguments();
+    if (triggerRuntimeArgs == null) {
+      if (!argumentMappings.isEmpty()) {
+        LOG.warn("No runtime arguments in the run '{}' of the triggering pipeline '{}' in namespace '{}' ",
+                 triggerInfo.getRunId(), triggerInfo.getApplicationSpecification().getName(),
+                 triggerInfo.getNamespace());
+      }
+      return;
+    }
+    for (TriggeringPipelinePropertyMapping.ArgumentMapping mapping : argumentMappings) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: This classname is incredibly long. ArgumentMapping should be its own class, not an inner class.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 24]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+        }
+        continue;
+      }
+      // Use the argument name in the triggering pipeline if target is not specified
+      String key = mapping.getTarget() == null ? mapping.getSource() : mapping.getTarget();
+      runtimeArgs.put(key, value);
+    }
+  }
+
+  private void addTriggeringPipelinePluginProperties(
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: there isn't really any value in having this as a separate method either.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 25]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+    List<TriggeringPipelinePropertyMapping.PluginPropertyMapping> pluginMapping,
+    ProgramStatusTriggerInfo triggerInfo, Map<String, String> runtimeArgs) {
+    Map<String, Plugin> plugins = triggerInfo.getApplicationSpecification().getPlugins();
+
+    for (TriggeringPipelinePropertyMapping.PluginPropertyMapping mapping : pluginMapping) {
+      Plugin plugin = plugins.get(mapping.getStageName());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: you need to handle the case when stageName is null due to bad user data. Skip that mapping if it's null.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 26]
```diff
@@ -296,22 +304,110 @@ protected void configure() {
...
+      if (plugin == null) {
+        LOG.warn("No plugin with name '{}' can be found in triggering pipeline '{}' in namespace '{}' ",
+                 mapping.getStageName(), triggerInfo.getApplicationSpecification().getName(),
+                 triggerInfo.getNamespace());
+        continue;
+      }
+      String value = plugin.getProperties().getProperties().get(mapping.getSource());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: similarly, check if source is null and skip if it is.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 27]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
...
+      .addConnection("filter1", "filter2")
+      .addConnection("filter2", "sink")
+      .setEngine(engine)
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("triggeredActionPipeline-" + engine);
+    ApplicationManager appManager = deployApplication(appId.toId(), appRequest);
+
+    // there should be only two programs - one workflow and one mapreduce/spark
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: this comment is unrelated to what happens below

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 28]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
   }
 
   @Test
-  public void testMacroActionPipelines() throws Exception {
+  public void testMacroActionPipelinesAndSchedule() throws Exception {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: don't hijack another unit test like this. Create your own that sets up two pipelines and has a logical flow. What you've done here is some weird mix that tests macros like before, but tests triggered stuff on one half of the pipelines with setup spread across multiple methods, making it difficult to tell what is being tested.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 29]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
...
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("triggeredActionPipeline-" + engine);
+    ApplicationManager appManager = deployApplication(appId.toId(), appRequest);
+
+    // there should be only two programs - one workflow and one mapreduce/spark
+    Schema schema = Schema.recordOf("testRecord", Schema.Field.of("name", Schema.of(Schema.Type.STRING)));
+    StructuredRecord recordSamuel = StructuredRecord.builder(schema).set("name", "samuel").build();
+    StructuredRecord recordKey1Value = StructuredRecord.builder(schema).set("name", "macroValue").build();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: this is not good, you have to read another method in order to understand why you have 'macroValue' here, and now your two methods are invisibly coupled.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 30]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
...
+    StructuredRecord recordKey2Value = StructuredRecord.builder(schema).set("name", "action1.row").build();
+
+    // write one record to each source
+    DataSetManager<Table> inputManager = getDataset(sourceName);
+    MockSource.writeInput(inputManager, ImmutableList.of(recordSamuel, recordKey1Value, recordKey2Value));
+
+    String defaultNamespace = NamespaceId.DEFAULT.getNamespace();
+    String triggeringPipeline = "macroActionTest-SPARK";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: same here. Just create a new test that creates two pipelines in the same method. There should be descriptive comments about what they do and what arguments and plugin properties from the first are expected to be passed to the second.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 31]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
...
+
+      // use the value of runtime argument with key "value" in triggering pipeline as the value for key1
+    propertyMapping.addArgumentMapping(new TriggeringPipelinePropertyMapping.ArgumentMapping("value", key1));
+      // use the value of property "rowKey" in plugin "action1" in triggering pipeline as the value for key2
+    propertyMapping.addPluginPropertyMapping(
+      new TriggeringPipelinePropertyMapping.PluginPropertyMapping("action1", "rowKey", key2));
+    ProgramStatusTrigger completeTrigger =
+      new ProgramStatusTrigger(new WorkflowId(defaultNamespace, triggeringPipeline, SmartWorkflow.NAME),
+                               ImmutableSet.of(ProgramStatus.COMPLETED));
+    ScheduleId scheduleId = appId.schedule(("completeSchedule"));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: redundant parentheses

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java [hunk 32]
```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
...
+
+  private void assertTriggeredPipelinesResult(WorkflowManager workflowManager, Engine engine) throws Exception {
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 3, TimeUnit.MINUTES);
+    List<RunRecord> runRecords = workflowManager.getHistory(ProgramRunStatus.COMPLETED);
+    Assert.assertEquals(1, runRecords.size());
+    String key1 = "trigger-runtime-arg";
+    String key2 = "trigger-plugin-property";
+    Map<String, List<WorkflowTokenDetail.NodeValueDetail>> tokenData =
+      workflowManager.getToken(runRecords.get(0).getPid(), null, null).getTokenData();
+    Assert.assertEquals("macroValue", tokenData.get(key1).get(0).getValue());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: the fact that it's in the workflow token is an implementation detail that you shouldn't be asserting. If we change the implementation, this test will fail. It is enough to check the data in the sink and data in the tables.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java (hunk 30 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java]: actually, your test should have 3 pipelines chained together. That way it would have caught the argument bug where it's not getting arguments from anything before the direct parent.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java (hunk 14 above)
Comment (ecfm) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: it should never be null. No need of this check
